pkg/controller/generic/qtransform/qtransform_test.go

```diff
@@ -278,6 +283,59 @@ func TestMapWithErrors(t *testing.T) {
 	})
 }
 
+func TestDestroyOutput(t *testing.T) {
+	setup(t, func(ctx context.Context, st state.State, runtime *runtime.Runtime) {
+		require.NoError(t, runtime.RegisterQController(NewABController(nil, nil)))
+
+		// prepare
+
+		res := NewA("1", ASpec{Str: "regular"})
+
+		require.NoError(t, st.Create(ctx, res))
+
+		rtestutils.AssertResources(ctx, t, st, []resource.ID{"transformed-1"}, func(r *B, assert *assert.Assertions) {
+			assert.Equal(`"regular"-0`, r.TypedSpec().Out)
+		})
+
+		// add a finalizer to the output
+		require.NoError(t, st.AddFinalizer(ctx, NewB("transformed-1", BSpec{}).Metadata(), "some-finalizer"))
+
+		// trigger the destroy-output
+		_, err := safe.StateUpdateWithConflicts(ctx, st, NewA("1", ASpec{}).Metadata(), func(a *A) error {
+			a.TypedSpec().Str = "destroy-output"
+
+			return nil
+		})
+		require.NoError(t, err)
+
+		rtestutils.AssertResource[*B](ctx, t, st, "transformed-1", func(r *B, assert *assert.Assertions) {
+			assert.Equal(resource.PhaseTearingDown, r.Metadata().Phase())
+		})
+
+		// update the input one more time
+		_, err = safe.StateUpdateWithConflicts(ctx, st, NewA("1", ASpec{}).Metadata(), func(a *A) error {
+			a.TypedSpec().Str = "regular2"
+
+			return nil
+		})
+		require.NoError(t, err)
+
+		// no changes expected
+		sleep(ctx, 250*time.Millisecond)
+
+		rtestutils.AssertResource[*B](ctx, t, st, "transformed-1", func(r *B, assert *assert.Assertions) {
+			assert.Equal(resource.PhaseTearingDown, r.Metadata().Phase())
+			assert.Equal(`"regular"-0`, r.TypedSpec().Out)
+		})
+
+		// remove the finalizer on the output
+		require.NoError(t, st.RemoveFinalizer(ctx, NewB("transformed-1", BSpec{}).Metadata(), "some-finalizer"))
+
+		// the output should be destroyed now
+		rtestutils.AssertNoResource[*B](ctx, t, st, "transformed-1")
```

Comment:
we should verify there that new resource got created?

Reply:
ah, I missed that part. Changed again, now I think it works as it should, but not sure about what I do to get it to requeue here: https://github.com/cosi-project/runtime/pull/452/files#diff-ad6500f6caec4671f4830d1c934e07a96c03689c01e0a65a1565ec2e0a279ffaR301

Reply:
I think in that case you need to return an error which will return before calling `WriterModify`, but it won't be an error from a controller point of view if finalizers are not empty

Reply:
done as discussed